src/syntax/binding_to_syntax.ts

```diff
@@ -6,7 +6,7 @@ import { BindingWhenOnSyntax } from "./binding_when_on_syntax";
 
 class BindingToSyntax<T> implements interfaces.BindingToSyntax<T> {
 
-  private _binding: interfaces.Binding<T>;
+  public _binding: interfaces.Binding<T>;
```

Comment:
For this, I think we should keep this member `private` and use the same tricks that are in use elsewhere to make it accessible with an `as unknown as { _binding: interfaces.Binding<T> }`, but only _for now_ :). Please see my following comment.

Reply:
I think that this is uncovering a smell.

I have noticed in the past that this member is `private`, but a type assertion is used in order to allow it to be accessed from a public location. This is a pattern that can be observed with other members, not just this one (for instance, `_bindingDictionary`).

A few places I have noticed this are among test files:

[`test/container/container.test.ts`](https://github.com/inversify/InversifyJS/blob/master/test/container/container.test.ts#L608)
[`test/syntax/binding_when_on_syntax.test.ts`](https://github.com/inversify/InversifyJS/blob/master/test/syntax/binding_when_on_syntax.test.ts#L32)

Another place is in [`src/planning/planner.ts`](https://github.com/inversify/InversifyJS/blob/master/src/planning/planner.ts#L21).

The reality of this smell is that there is a need for two forms of the containing types (in this case, I am referring to `BindingToSyntax` since this is where we have concentrated this specific example):

1. A _presentational_ version which maintains a correct API for the consumers of the package, and does not expose unnecessary members like `_binding`.
2. An _internal_ version which allows a more exposed API for the purposes of testing and proper layer integration with `planner` and friends.

I think that a good way of accomplishing this which resolves the smell is to rename the class with a `_` prefix, as in `BindingToSyntax => _BindingToSyntax`, etc. This means that the current definition of this class will become the internal form of the class. Then, any members which are meant to be public may be accomplished in the following manner:

```typescript
// Please ignore the lack of generic type vars, etc.,
// as this is only meant to be an example
class BindingToSyntax {
  constructor(
    private _internal = new _BindingToSyntax(...args)
  ) {}

  get somePublicMemberFromOriginalImplementation() {
    return this._internal.somePublicMemberFromOriginalImplementation;
  }
}
```

etc.

Finally, with a presentational/internal pattern like this, we are free to fix the original problem for those classes which have the behavior of needing a private member exposed for testing, or for use in `planner` or similar. We can simply make the members that needed this special access despite being `private` members into `public` ones instead.

What do you think?

Reply:
If you agree @PodaruDragos then I will spin the above comment off as an issue for us to work on at some time.

Reply:
totally agree with this. Do you also agree that we should just move everything to actual `#private` fields since you know they are really `private` ?

Reply:
I think it is worth taking a look at cleaning this up

Reply:
@PodaruDragos I think we should definitely do `#private` type of vars in the future, for now let's not increase the scope of this PR